### DISCUSSION TITLE
tryOnScopeDispose() Explained

### DISCUSSION THREAD

**Discussion by michealroberts** [vueuse/vueuse#2513] [Q&A]
Could someone far more intelligent than me explain why we need the `if (getCurrentScope()) {}` check in tryOnScopeDispose, surely, `onScopeDispose` will only be called if the ... current scope is disposed.

I've had some really great discussions with a colleague and for us, we're convinced that it is not necessary and the warning should be shown to the developer.

Essentially, I don't think we should be bypassing the core warning ...

cc @antfu

**Comment by sibbng** [comment]
You can use VueUse utilities outside of a scope/component. If you call `onScopeDispose` without a scope, it will show a warning on console during development. We need that check to bypass the warning.

https://github.com/vuejs/core/blob/fe77e2bddaa5930ad37a43fe8e6254ddb0f9c2d7/packages/reactivity/src/effectScope.ts#L127-L136

**Comment by michealroberts** [comment]
@sibbng Could you outline an example scenario where we don't have a scope? Surely the warning is in place for a reason, we shouldn't be calling composables outside of scope?

**Reply by sibbng** [reply]
If you don't initialize composable inside a Vue component or don't use [effectScope ](https://vuejs.org/api/reactivity-advanced.html#effectscope) when initializing outside a Vue component you don't have a scope. This is useful in two scenarios:

1. Global state
If you are working on an SPA, it may be useful to initialize your composables outside of a component to share the state across components. Don't do that in SSR frameworks like Nuxt. Because the state will be shared across requests and often you wouldn't want that. And, if I remember correctly those variables will not be shown in Devtools. 

2. Usage with other frameworks
Not forcing you to have a scope, giving you full control over the lifetime of composable. Most VueUse utilities return a stop function that stops the attached event listeners and watchers. This is especially useful when working with other frameworks. Here's an example implementation of onClickOutside with Solid:
https://playground.solidjs.com/anonymous/8793ce4d-d740-457b-90f4-d4ff23fccf32
You can do the same with effectScope too but that will result in too much boilerplate code and overhead in the memory for such a simple use case.
